### PR TITLE
map retention with policy

### PR DESCRIPTION
**Pull request by bitsf**

### CODE REVIEW

File: src/pkg/retention/controller.go [hunk 1]
```diff
@@ -27,4 +38,212 @@ type APIController interface {
...
+		return err
+	}
+	if p.Trigger.Kind == "Schedule" {
+		err = r.scheduler.UnSchedule(p.Trigger.References["jobid"].(string))
+		if err != nil {
+			return err
+		}
+	}
+
+	execs, err := r.manager.ListExecutions(id, nil)
```
Comment (steven-zou) [inline src/pkg/retention/controller.go]: move to the DAO layer

File: src/pkg/retention/controller.go [hunk 2]
```diff
@@ -27,4 +38,212 @@ type APIController interface {
...
+		err = r.manager.DeleteExecutions(e.ID, nil)
+		return err
+	}
+
+	return r.manager.DeletePolicy(id)
+}
+
+// TriggerRetentionExec Trigger Retention Execution
+func (r *DefaultAPIController) TriggerRetentionExec(policyID int64, trigger string) error {
+	launcher := NewLauncher(r.projectManagerNew, r.repositoryMgr, r.manager, dep.DefaultClient)
```
Comment (steven-zou) [inline src/pkg/retention/controller.go]: move to a ref of controller

Comment (steven-zou) [reply src/pkg/retention/controller.go]:
```
New(client dep.Client) Controller
```

Comment (steven-zou) [reply src/pkg/retention/controller.go]: `api.Init`

File: src/core/api/retention.go [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+{
+    "templates": [
+        {
+            "rule_template": "lastXDays",
+            "display_text": "last X days",
+            "action": "retain",
+            "params": [
+                {
+                    "name": "lastXDays",
+                    "display_text": "last X days",
```
Comment (steven-zou) [inline src/core/api/retention.go]: remove `display_text ` of param

File: src/core/api/retention.go [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+}
+
+// GetMetadatas Get Metadatas
+func (r *RetentionAPI) GetMetadatas() {
+	data := `
+{
+    "templates": [
+        {
+            "rule_template": "lastXDays",
+            "display_text": "last X days",
```
Comment (steven-zou) [inline src/core/api/retention.go]: "last X days" => the images from the last # days

File: src/core/api/retention.go [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+                    "display_text": "latest X images",
+                    "type": "int",
+                    "unit": "count",
+                    "required": true
+                }
+            ]
+        },
+        {
+            "rule_template": "latestPulledK",
+            "display_text": "latest pulled X images",
```
Comment (steven-zou) [inline src/core/api/retention.go]: =>the most recently pulled # images

File: src/core/api/retention.go [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+                    "display_text": "latest active X images",
+                    "type": "int",
+                    "unit": "count",
+                    "required": true
+                }
+            ]
+        },
+        {
+            "rule_template": "latestK",
+            "display_text": "latest X images",
```
Comment (steven-zou) [inline src/core/api/retention.go]: =>the most recently pushed # images

File: src/core/api/retention.go [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+            ]
+        },
+        {
+            "rule_template": "latestActiveK",
+            "display_text": "latest active X images",
+            "action": "retain",
+            "params": [
+                {
+                    "name": "latestActiveK",
+                    "display_text": "latest active X images",
```
Comment (steven-zou) [inline src/core/api/retention.go]: =>the most recent active # images

File: src/pkg/retention/controller.go [hunk 8]
```diff
@@ -27,4 +37,220 @@ type APIController interface {
...
+func (r *DefaultAPIController) GetRetention(id int64) (*policy.Metadata, error) {
+	return r.manager.GetPolicy(id)
+}
+
+// CreateRetention Create Retention
+func (r *DefaultAPIController) CreateRetention(p *policy.Metadata) error {
+	if p.Scope.Level != "project" {
+		return fmt.Errorf("scope %s is not support", p.Scope.Level)
+	}
+	if p.Scope.Reference <= 0 {
```
Comment (steven-zou) [inline src/pkg/retention/controller.go]: this check only happens when the scope is `project`

File: src/pkg/retention/manager.go [hunk 9]
```diff
@@ -212,38 +217,15 @@ func (d *DefaultManager) GetTaskLog(taskID int64) ([]byte, error) {
...
-		h1.Artifact = h.Artifact
-		h1.Rule.ID = h.RuleID
-		h1.Rule.DisplayText = h.RuleDisplayText
-		h1.Timestamp = h.Timestamp
-		his1 = append(his1, h1)
-	}
-	return his1, nil
+// DeleteExecutions Delete Executions
+func (d *DefaultManager) DeleteExecutions(policyID int64, query *q.Query) error {
+	panic("implement me")
```
Comment (steven-zou) [inline src/pkg/retention/manager.go]: why it is still not implemented till now?

File: src/pkg/retention/controller.go (hunk 8 above)
Comment (steven-zou) [reply src/pkg/retention/controller.go]:
```go
switch p.Scope.Level {
  case "project":
     if p.Scope.Reference <= 0 {}
  default:
   return fmt.Errorf("scope %s is not support", p.Scope.Level)
}

```

File: src/pkg/retention/manager.go (hunk 9 above)
Comment (steven-zou) [reply src/pkg/retention/manager.go]: remove it if it is not used anymore

File: src/core/api/retention.go [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+		return
+	}
+	r.Data["json"] = p
+	r.ServeJSON()
+}
+
+// CreateRetention Create Retention
+func (r *RetentionAPI) CreateRetention() {
+	p := &policy.Metadata{}
+	isValid, err := r.DecodeJSONReqAndValidate(p)
```
[the remark this answers is not in the document]
Comment (bitsf) [reply src/core/api/retention.go]: will use another PR to implement this

File: src/pkg/retention/controller.go [hunk 11]
```diff
@@ -27,4 +37,237 @@ type APIController interface {
...
+
+	GetRetention(id int64) (*policy.Metadata, error)
+
+	CreateRetention(p *policy.Metadata) error
+
+	UpdateRetention(p *policy.Metadata) error
+
+	DeleteRetention(id int64) error
+
+	TriggerRetentionExec(policyID int64, trigger string, idDryRun bool) error
```
Comment (steven-zou) [inline src/pkg/retention/controller.go]: idDryRun =>`dryRun`?

File: src/pkg/retention/controller.go [hunk 12]
```diff
@@ -27,4 +37,237 @@ type APIController interface {
...
+		}
+		proj, err := r.projectManager.Get(p.Scope.Reference)
+		if err != nil {
+			return (err)
+		}
+		if proj == nil {
+			return fmt.Errorf("Invalid Project id %d", p.Scope.Reference)
+		}
+
+		if p.Trigger.Kind == "Schedule" {
```
Comment (steven-zou) [inline src/pkg/retention/controller.go]: The code below should be sharable code for all scopes, right?

File: src/core/api/base.go [hunk 13]
```diff
@@ -39,8 +41,11 @@ const (
 
 // the managers/controllers used globally
 var (
-	projectMgr    project.Manager
-	repositoryMgr repository.Manager
+	projectMgr         project.Manager
+	repositoryMgr      repository.Manager
+	retentionScheduler retention.Scheduler
```
Comment (ywk253100) [inline src/core/api/base.go]: There is no retention scheduler anymore, use the pkg/scheduler instead: https://github.com/goharbor/harbor/blob/feature/tag_retention/src/pkg/scheduler/scheduler.go#L40

**Pull request merged**